src/mykrobe/_vcf/models.py

```diff
@@ -159,79 +157,182 @@ def global_variant_set(self):
+                    v += c  # include quote mark in value
+                    state = 1  # end quoting
+                else:
+                    v += c
+        if k != "":
+            val[k] = v
+        return key, val
+
+    def _read_meta(self, meta_string: str) -> Tuple[str, Union[str, Dict[str, str]]]:
+        """Taken from pyvcf
+        https://github.com/jamescasbon/PyVCF/blob/476169cd457ba0caa6b998b301a4d91e975251d9/vcf/parser.py#L222-L231
+        """
+        if re.match("##.+=<", meta_string):
+            return self._read_meta_hash(meta_string)
+        meta_pattern = re.compile(r"""##(?P<key>.+?)=(?P<val>.+)""")
+        match = meta_pattern.match(meta_string)
+        if not match:
+            # Spec only allows key=value, but we try to be liberal and
+            # interpret anything else as key=none (and all values are parsed
+            # as strings).
+            return meta_string.lstrip("#"), "none"
+        return match.group("key"), match.group("val")
+
+    def _metadata_from_header(
+        self, raw_header: str
+    ) -> Dict[str, Union[str, List[str]]]:
+        metadata = dict()
+        for line in raw_header.splitlines():
+            key, val = self._read_meta(line)
+            if key in NON_METADATA_KEYS:
+                continue
+            if key in SINGULAR_METADATA:
+                metadata[key] = val
+            else:
+                if key not in metadata:
+                    metadata[key] = []
+                metadata[key].append(val)
+
+        return metadata
+
+    @staticmethod
+    def _is_record_valid(record: cyvcf2.Variant) -> bool:
+        # todo: review this method. There are some confusing assumptions about the number of alts we expect and whether HETs are valid or not
         valid = True
-        for sample in record.samples:
-            if sample["GT"] is None:
+        for i, gt in enumerate(record.genotypes):
+            if record.gt_types[i] == 3:  # null (unknown) genotype
                 valid = False
             else:
-                try:
-                    if sum([int(i) for i in split_GT(sample['GT'])]) < 2:
-                        valid = False
-                except ValueError:
+                alleles = [a for a in gt if type(a) is int]
+                if len(alleles) < 2:
+                    # we add 1 here in case the GT field is haploid and has a GT of 1
+                    # which is analogous to 1/1 in a diploid. Cortex produces diploid
+                    # calls, so this is just for completeness
+                    alleles.append(1)
+                if any(
+                    a < 0 for a in alleles
+                ):  # todo: this copies the previous implementation, we basically say anything that has a null GT '.', even if HET is invalid
+                    valid = False
+                if (
+                    sum(alleles) < 2
+                ):  # todo: is the goal here to exclude HETs? If so, this will fail with a GT of 0/2
```

Comment:
yeah, good catch. I expect the intention was to exclude hets?

We could do something like `if len(set(alleles)) > 1`?